source/fe/mapping_q_cache.cc

```diff
@@ -249,17 +259,465 @@ MappingQCache<dim, spacedim>::initialize(
+      &vector_ghosted)
+  {
+    for (const auto i : vector_ghosted.locally_owned_elements())
+      vector_ghosted[i] = vector[i];
+  }
+
+  template <typename VectorType,
+            typename std::enable_if<
+              std::is_same<VectorType,
+                           LinearAlgebra::distributed::Vector<
+                             typename VectorType::value_type>>::value,
+              VectorType>::type * = nullptr>
+  void
+  copy_locally_owned_data_from(
+    const VectorType &vector,
+    LinearAlgebra::distributed::Vector<typename VectorType::value_type>
+      &vector_ghosted)
+  {
+    vector_ghosted.copy_locally_owned_data_from(vector);
+  }
+
+#if defined(DEAL_II_WITH_TRILINOS) && defined(DEAL_II_WITH_MPI)
+  void
+  copy_locally_owned_data_from(
+    const LinearAlgebra::EpetraWrappers::Vector &vector,
+    LinearAlgebra::distributed::Vector<
+      typename LinearAlgebra::EpetraWrappers::Vector::value_type>
+      &vector_ghosted)
+  {
+    Assert(false, ExcNotImplemented());
+    (void)vector;
+    (void)vector_ghosted;
+  }
+#endif
+} // namespace
+
+
+
+template <int dim, int spacedim>
+template <typename VectorType>
+void
+MappingQCache<dim, spacedim>::initialize(
+  const Mapping<dim, spacedim> &   mapping,
+  const DoFHandler<dim, spacedim> &dof_handler,
+  const VectorType &               vector,
+  const bool                       vector_describes_relative_displacement)
+{
+  AssertDimension(dof_handler.get_fe_collection().size(), 1);
+  const FiniteElement<dim, spacedim> &fe = dof_handler.get_fe();
+  AssertDimension(fe.n_base_elements(), 1);
+  AssertDimension(fe.element_multiplicity(0), spacedim);
+
+  const unsigned int is_fe_q =
+    dynamic_cast<const FE_Q<dim, spacedim> *>(&fe.base_element(0)) != nullptr;
+  const unsigned int is_fe_dgq =
+    dynamic_cast<const FE_DGQ<dim, spacedim> *>(&fe.base_element(0)) != nullptr;
+
+  const auto lexicographic_to_hierarchic_numbering =
+    Utilities::invert_permutation(
+      FETools::hierarchic_to_lexicographic_numbering<spacedim>(
+        this->get_degree()));
+
+  // Step 1: copy global vector so that the ghost values are such that the
+  // cache can be set up for all ghost cells
+  LinearAlgebra::distributed::Vector<typename VectorType::value_type>
+           vector_ghosted;
+  IndexSet locally_relevant_dofs;
+  DoFTools::extract_locally_relevant_dofs(dof_handler, locally_relevant_dofs);
+  vector_ghosted.reinit(dof_handler.locally_owned_dofs(),
+                        locally_relevant_dofs,
+                        dof_handler.get_communicator());
+  copy_locally_owned_data_from(vector, vector_ghosted);
+  vector_ghosted.update_ghost_values();
+
+  // FE and FEValues in the case they are needed
+  FE_Nothing<dim, spacedim> fe_nothing;
+  Threads::ThreadLocalStorage<std::unique_ptr<FEValues<dim, spacedim>>>
+             fe_values_all;
+  const bool update_values_might_be_needed =
+    ((is_fe_q || is_fe_dgq) && fe.degree == this->get_degree()) == false;
```

Comment:
This variable name could be improved - don't we always need `update_values` if we cannot just read off locations from the solution vectors?

Reply:
I have renamed the variable to `interpolation_of_values_is_needed` and have added a comment. I now also use this variable at more places (this variable was only introduced later^^).